spec/unit/room.spec.ts

```diff
@@ -3147,106 +3148,195 @@ describe("Room", function () {
+                // Given a threaded receipt exists, in main thread
+                room.getReadReceiptForUserId = (): WrappedReceipt | null => {
+                    return { eventId: "rootId", data: { ts: 12, thread_id: "main" } } as WrappedReceipt;
+                };
+                // And the event it refers to is in a thread, because it is a thread root
+                room.findEventById = jest
+                    .fn()
+                    .mockReturnValue({ isThreadRoot: true, threadRootId: "thread1" } as MatrixEvent);
 
-                    expect(room.getEventReadUpTo(userA)).toEqual(`eventId2`);
-                });
+                // When we ask what they have read
+                // Then we say the event
+                expect(room.getEventReadUpTo(userA)).toEqual("rootId");
             });
+        });
 
-            describe("fallback precedence", () => {
-                beforeAll(() => {
-                    room.getUnfilteredTimelineSet = () =>
-                        ({
-                            compareEventOrdering: (_1, _2) => null,
-                        } as EventTimelineSet);
-                });
+        describe("valid receipts", () => {
+            beforeEach(() => {
+                // When we look up the event referred to by the receipt, it exists
+                room.findEventById = jest.fn().mockReturnValue({} as MatrixEvent);
+            });
 
-                it("should give precedence to m.read.private", () => {
+            it("handles missing receipt type", () => {
+                room.getReadReceiptForUserId = (userId, ignore, receiptType): WrappedReceipt | null => {
+                    return receiptType === ReceiptType.ReadPrivate ? ({ eventId: "eventId" } as WrappedReceipt) : null;
+                };
+                expect(room.getEventReadUpTo(userA)).toEqual("eventId");
+            });
+
+            describe("prefers newer receipt", () => {
+                it("should compare correctly using timelines", () => {
                     room.getReadReceiptForUserId = (userId, ignore, receiptType): WrappedReceipt | null => {
                         if (receiptType === ReceiptType.ReadPrivate) {
-                            return { eventId: "eventId1", data: { ts: 123 } };
+                            return { eventId: "eventId1" } as WrappedReceipt;
                         }
                         if (receiptType === ReceiptType.Read) {
-                            return { eventId: "eventId2", data: { ts: 123 } };
+                            return { eventId: "eventId2" } as WrappedReceipt;
                         }
                         return null;
                     };
 
-                    expect(room.getEventReadUpTo(userA)).toEqual(`eventId1`);
+                    for (let i = 1; i <= 2; i++) {
+                        room.getUnfilteredTimelineSet = () =>
+                            ({
+                                compareEventOrdering: (event1: string) => {
```

Comment:
Nit: I'm unconvinced that dropping `event2` here is an improvement in clarity. It makes me want to dig out the definition of `compareEventOrdering` to see how it can possibly compare the ordering of a single event.

Better to just give it a name which reflects the fact we don't need it here:

```suggestion
                                compareEventOrdering: (event1: string, _event2: string) => {
```